File: 1311-largest-magic-square/largest-magic-square.java
class Solution {
    public int largestMagicSquare(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] rowPrefix = new int[m][n + 1];
        int[][] colPrefix = new int[n][m + 1];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                rowPrefix[i][j + 1] = rowPrefix[i][j] + grid[i][j];
                colPrefix[j][i + 1] = colPrefix[j][i] + grid[i][j];
            }
        }

        for (int k = Math.min(m, n); k > 1; k--) {
            for (int r = 0; r <= m - k; r++) {
                for (int c = 0; c <= n - k; c++) {
                    if (isMagic(grid, rowPrefix, colPrefix, r, c, k)) {
                        return k;
                    }
                }
            }
        }
        return 1;
    }
private boolean isMagic(int[][] grid, int[][] rowPrefix, int[][] colPrefix, int r, int c, int k) {
        Integer target = null;
        for (int i = r; i < r + k; i++) {
            int sum = rowPrefix[i][c + k] - rowPrefix[i][c];
            if (target == null) target = sum;
            else if (target != sum) return false;
        }
        for (int j = c; j < c + k; j++) {
            int sum = colPrefix[j][r + k] - colPrefix[j][r];
            if (target != sum) return false;
        }
        int diag1 = 0, diag2 = 0;
        for (int i = 0; i < k; i++) {
            diag1 += grid[r + i][c + i];
            diag2 += grid[r + i][c + k - 1 - i];
        }
        return target == diag1 && target == diag2;
    }
}

File: 1311-largest-magic-square/largest-magic-square.py
from typing import List

class Solution:
    def largestMagicSquare(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        row_prefix = [[0]*(n+1) for _ in range(m)]
        col_prefix = [[0]*(m+1) for _ in range(n)]
        
        for i in range(m):
            for j in range(n):
                row_prefix[i][j+1] = row_prefix[i][j] + grid[i][j]
                col_prefix[j][i+1] = col_prefix[j][i] + grid[i][j]
        
        def is_magic(r, c, k):
            target = None
            for i in range(r, r+k):
                s = row_prefix[i][c+k] - row_prefix[i][c]
                if target is None:
                    target = s
                elif target != s:
                    return False
            for j in range(c, c+k):
                s = col_prefix[j][r+k] - col_prefix[j][r]
                if target != s:
                    return False
            diag1 = sum(grid[r+i][c+i] for i in range(k))
            if target != diag1:
                return False
            diag2 = sum(grid[r+i][c+k-1-i] for i in range(k))
            if target != diag2:
                return False
            return True
        for k in range(min(m, n), 1, -1):
            for r in range(m-k+1):
                for c in range(n-k+1):
                    if is_magic(r, c, k):
                         return k
        return 1

